### ISSUE TITLE
Pipewire: Support modifying sink volume instead of stream volume

### ISSUE THREAD

**Issue by MagicMuscleMan** [MusicPlayerDaemon/MPD#1588]
When using Pipewire as an output plugin and changing the volume in MPD the stream volume is modified. If the computer is both used as a music player and for other things, this is problematic.

Either the general volume (sink volume) is on a normal level (e.g. 10%), then MPD is limited to 100% stream volume and there is no possibility to listen to music a bit louder than when sitting in front of the computer (overall (obviously simplified) 10% * 100% = 10%).

Alternatively, the sink volume is loud per default (e.g. 100%) and all used applications have a low stream value (e.g. 10%), then MPD has the full dynamic range. However, as soon as a new application outputs some sounds, all neighbors are informed about that, as the (new) stream value default is 100% (resulting in 100% * 100% = 100%). ;-)

Therefore, there should be an option to switch between controlling sink volume and stream volume when using Pipewire.

**Comment by MagicMuscleMan** [comment]
I wrote a patch how this feature could be implemented. If MPD's config file contains a Pipewire audio_output section with the entry `mixer_volume "sink"`, then the sink volume is used and modified. For `mixer_volume "stream"` (and as default value) the behavior is as before.

However, I guess this patch is not mergeable, as

1. proper error handling is missing
2. test depth is low
3. it uses a system call to read and set the volume
4. it depends on Wireplumber being used

Solving 1. would be possible, but makes most sense after 3. and 4. is solved. I will implicitly work on 2. during the next days (I just put the code on the living room's MPD). However, 3. seems to be the tricky one. Pipewire's library documentation is hardly comprehensible, the source of the currently used tool wpctl is quite complex and Wireplumber specific.

So the most promising way forward would be to follow the [pw-cli example](https://gitlab.freedesktop.org/pipewire/pipewire/-/wikis/Migrate-PulseAudio#sinksource-port-volumemuteport-latency) in pw-cli's source code and call the necessary library functions from MPD. However, I will probably not work on that due to other tasks.

If anyone wants to continue, here is the patch:
```
diff '--color=auto' -ruN mpd-0.23.8-old/src/mixer/plugins/PipeWireMixerPlugin.cxx mpd-0.23.8-new/src/mixer/plugins/PipeWireMixerPlugin.cxx
--- mpd-0.23.8-old/src/mixer/plugins/PipeWireMixerPlugin.cxx	2022-07-09 01:05:38.000000000 +0200
+++ mpd-0.23.8-new/src/mixer/plugins/PipeWireMixerPlugin.cxx	2022-08-09 17:07:44.000000000 +0200
@@ -24,17 +24,32 @@

 #include <cmath>

+#include "config/Block.hxx"
+#include <cstring>
+#include <memory>
+// #include "Log.hxx"
+// #include "util/Domain.hxx"
+
+// static constexpr Domain pipewire_mixer_domain("pipewire_mixer");
+
 class PipeWireMixer final : public Mixer {
 	PipeWireOutput &output;

 	int volume = 100;

+	bool sink_volume = false;
+
 public:
 	PipeWireMixer(PipeWireOutput &_output,
-		      MixerListener &_listener) noexcept
+		      MixerListener &_listener,
+			  const ConfigBlock &block) noexcept
 		:Mixer(pipewire_mixer_plugin, _listener),
 		 output(_output)
 	{
+		const char *const volume_name(
+			block.GetBlockValue("mixer_volume", nullptr));
+		if (volume_name != nullptr)
+			sink_volume = !strcmp(volume_name, "sink");
 	}

 	~PipeWireMixer() noexcept override;
@@ -65,27 +80,60 @@
 	pm.OnVolumeChanged(new_volume);
 }

+static std::string
+exec(const char* cmd) {
+    std::array<char, 128> buffer;
+    std::string result;
+    std::unique_ptr<FILE, decltype(&pclose)> pipe(popen(cmd, "r"), pclose);
+    if (!pipe) {
+        throw std::runtime_error("popen() failed!");
+    }
+    while (fgets(buffer.data(), buffer.size(), pipe.get()) != nullptr) {
+        result += buffer.data();
+    }
+    return result;
+}
+
 int
 PipeWireMixer::GetVolume()
 {
+	/* If global volume is active, use the sink's volume internally. */
+	if (sink_volume) {
+		// FmtWarning(pipewire_mixer_domain, "Old mixer volume: {}", volume);
+		std::string vol = exec("/bin/wpctl get-volume @DEFAULT_AUDIO_SINK@");
+		// vol contains e.g. "Volume: 0.78" whereas e.g. "78" is needed
+		size_t index_first_digit = vol.find_first_of(' ') + 1;
+		volume = round(std::stof(vol.substr(index_first_digit)) * 100.f);
+		// FmtWarning(pipewire_mixer_domain, "New volume of mixer from default audio sink: {}", volume);
+	}
+
 	return volume;
 }

 void
 PipeWireMixer::SetVolume(unsigned new_volume)
 {
-	pipewire_output_set_volume(output, float(new_volume) * 0.01f);
+	if (sink_volume) {
+		// FmtWarning(pipewire_mixer_domain, "In SetVolume: {}", float(new_volume) * 0.01f);
+
+		std::string setCmd("/bin/wpctl set-volume @DEFAULT_AUDIO_SINK@ ");
+		setCmd = setCmd + std::to_string(float(new_volume) * 0.01f);
+		std::string ret = exec(setCmd.c_str());
+	} else
+		pipewire_output_set_volume(output, float(new_volume) * 0.01f);
+
 	volume = new_volume;
 }

 static Mixer *
 pipewire_mixer_init([[maybe_unused]] EventLoop &event_loop, AudioOutput &ao,
 		    MixerListener &listener,
-		    const ConfigBlock &)
+		    const ConfigBlock &block)
 {
 	auto &po = (PipeWireOutput &)ao;
-	auto *pm = new PipeWireMixer(po, listener);
+	auto *pm = new PipeWireMixer(po, listener, block);
 	pipewire_output_set_mixer(po, *pm);
+
 	return pm;
 }

@@ -98,3 +146,24 @@
 	pipewire_mixer_init,
 	true,
 };
diff '--color=auto' -ruN mpd-0.23.8-old/src/mixer/plugins/PipeWireMixerPlugin.hxx mpd-0.23.8-new/src/mixer/plugins/PipeWireMixerPlugin.hxx
--- mpd-0.23.8-old/src/mixer/plugins/PipeWireMixerPlugin.hxx	2022-07-09 01:05:38.000000000 +0200
+++ mpd-0.23.8-new/src/mixer/plugins/PipeWireMixerPlugin.hxx	2022-08-09 07:36:10.000000000 +0200
@@ -20,6 +20,8 @@
 #ifndef MPD_PIPEWIRE_MIXER_PLUGIN_HXX
 #define MPD_PIPEWIRE_MIXER_PLUGIN_HXX

+#include <string>
+
 struct MixerPlugin;
 class PipeWireMixer;

@@ -28,4 +30,7 @@
 void
 pipewire_mixer_on_change(PipeWireMixer &pm, float new_volume) noexcept;

+static std::string
+exec(const char* cmd);
+
 #endif

```

**Comment by MaxKellermann** [comment]
Calling an external program is not acceptable. If you ever find out how to do it with a library call instead, we can talk about merging.

**Comment by gearhead** [comment]
I have a running patch to the current git which used the internal mpd volume to control the volume control on the pipewire/wireplumber device. This is done in cli via 'wpctl set-volume <node> #' and 'wpctl get-volume <node>'
With this patch, I am able to use the mpd volume control via mpc to actually control the hardware volume control onmy i2s DAC.
(I apologize, but this patch has an extra patched file in it... mpd.service.in...)

[wpctl_code.patch](https://github.com/user-attachments/files/29169348/wpctl_code.patch)

[wpctl_adds.patch](https://github.com/user-attachments/files/29169350/wpctl_adds.patch)

Apply these against the current base and they seem to work. 

the audio output config is:
After applying, rebuild, then:

  audio_output {
      name       "pipewire"
      type       "pipewire"
      mixer_type "wpctl"
      enabled    "yes"
  }

edit - there seems to be some disconnect this morning. I am still working on it, 
edit 2 - no changes. What I was seeing was that I had adjusted the stream volume in wire plumber and it caused the perceived volume to be low. I set it back to 1 and all was as expected. 

in summary, this patch does what I want it to do. I do not know if this is what you would like mpd to do or not. 

@MaxKellermann if more clarification is needed, let me know.

Gearhead

**Comment by gearhead** [comment]
I re-read this issue and have refashioned my patch to do, I think, what the poster wants. 
This patch applies cleanly to the latest git. (7b0ff8cc8). I have verified on my setup that it does adjust the volume of the stream or sink as commanded by the 'mixer_type' parameter in the audio_output section of mpd.conf:
audio_output {
	name 		"pipewire"
	type 		"pipewire"
	mixer_type 	"pwsink"
	enabled 	"yes"
}
This will adjust the Pipewire sink volume from the mpd interface. If mixer_type is set to "pwstream" it will control the stream volume.

Gearhead

[pipewire_volume.patch](https://github.com/user-attachments/files/29272292/pipewire_volume.patch)

**Comment by gearhead** [comment]
@MaxKellermann 
I think this patch solves the user's issue. Does it meet your request to do it with a library call? I do not know how to issue a pull request for this, maybe I cannot due to not being on the maintainers list. Please use this patch it as you see fit. 
My take is it allows for hardware volume control when using the pipewire backend but does not require it. if 'sink' is set as the mixer and pipewire provides the information on a hardware volume control (via ucm), the volume control does control the digital volume control on the alsa card.

**Comment by MagicMuscleMan** [comment]
@gearhead: I haven't checked the patch as I am busy with other topics for the time being, but it sounds promising. Thanks for working on it!

**Comment by borine** [comment]
I don't understand. Max said:
> Calling an external program is not acceptable

and your patch calls:
```c++
execv("/usr/bin/wpctl", const_cast<char *const *>(argv));
```
If that is not calling an external program then I don't know what is.

There are also some other issues here that need addressing, possibly by writing clearer documentation or else by generalizing the design.

The difficulty I see is that routing of audio channels through `pipewire` is dynamic and can be re-routed by any client at any time, much like`jack`. So when `mpd` requests a specific sink, or asks `pipewire` to connect one of of its choice, then that connection should be considered temporary and liable to change. In general it is not even guaranteed that all channels of a stream from `mpd` will be routed to the same physical sound card. Your model is good enough for a simple music player system with one soundcard but breaks down in a more complex setup where audio routing is managed through a "patchbay" application.

Also, `wpctl` does not allow to specify a node by name, only by numeric ID, and those IDs are assigned dynamically so you can't write it into the `mpd` config as a constant. But, as invoking `wpctl` is not allowed any way, I guess that issue will go away when someone codes to either the pipewire or wireplumber client library API.

**Comment by gearhead** [comment]
I will take that under advisement and return to the drawing board. Thanks for looking. 
I was unable to find a clear way of getting mpd to control the sink volume on the default output. Controlling the stream is, I guess, what mpd already does when it adjusts its output volume with a pipewire backend. It seemed to me to be a shame that we can do it with Alsa, but not with pipewire.

**Comment by borine** [comment]
> It seemed to me to be a shame that we can do it with Alsa, but not with pipewire.

ALSA does not re-route a running stream, its stream endpoints are always fixed. So this problem does not arise when sending output directly to an ALSA PCM. Pipewire, and Jack, sit on top of ALSA to provide this additional routing functionality together with the additional complexity that goes with it.

I think the original feature request is still valid, but requires a more refined specification to make clear the constraints it imposes on what a user can and cannot do with pipewire. These constraints then need to be made explicit in the documentation.

I suspect (but have no evidence) that `mpd` is very often used on systems where only one pipewire sink is active at a time, and the pipewire "default" can be manipulated such that it always references that one active sink. In these cases then directing volume change requests to the control(s) on the Device node(s) associated with whatever sink node is active at that time would have the desired effect.

**Comment by gearhead** [comment]
@borine
Went back to the drawing board and tried again. This patch does not call wpctl. It uses the pipewire protocol to vary sink volume when selected. In addition, I added the ability to specify a particular pipewire sink instead of 'default' if desired. This applies and builds to the latest git as of today 07039e1cc. tested with default (pwsink_target unset) and with the `node.name`, pwsink and pwstream. All seems to work. 

Let me know if this is more acceptable. 

In the config for audio output, these are the changes needed:
```
audio_output {
	name 		"pipewire"
	type 		"pipewire"
	mixer_type 	"pwsink"
	enabled      "yes"
}
```

mixer_type 'pwsink' controls the sink volume from mpc and 'pwstream' controls the stream volume.

If the user sets mixer_type as "pwsink' it will use the default card. If the user wants a different card, that can also be specified.
add a parameter 'pwsink_target' and set it equal to the node.name property from pipewire. The node.name can be found by querying pipewire. Either `wpctl status` or `pw-cli ls Node`

On my particular setup it is:
`pwsink_target	"alsa_output.platform-soc_sound.HiFi__Speaker__sink"`

The `node.name` is was found like this in the 'Audio/Sink line (I am logged in as user mpd):
```
$ wpctl status
PipeWire 'pipewire-0' [1.4.2, mpd@desktop, cookie:1514305422]
 └─ Clients:
        34. WirePlumber                         [1.4.2, mpd@desktop, pid:626]
        35. pipewire                            [1.4.2, mpd@desktop, pid:627]
        48. WirePlumber [export]                [1.4.2, mpd@desktop, pid:626]
        62. mpd-pwsink-mixer                    [1.4.2, mpd@desktop, pid:632]
        63. Music Player Daemon                 [1.4.2, mpd@desktop, pid:632]
        69. wpctl                               [1.4.2, mpd@desktop, pid:12493]

...

Settings
 └─ Default Configured Devices:
         0. Audio/Sink    alsa_output.platform-soc_sound.HiFi__Speaker__sink
```
Another way to find its `node.name` is using:
```
pw-cli ls Node
...
	id 58, type PipeWire:Interface:Node/3
 		object.serial = "58"
 		object.path = "alsa:acp:IQaudIODAC:0:playback"
 		factory.id = "19"
 		client.id = "48"
 		device.id = "50"
 		priority.session = "1000"
 		priority.driver = "1000"
 		node.description = "Built-in Audio IQaudIODAC output"
 		node.name = "alsa_output.platform-soc_sound.HiFi__Speaker__sink"
 		node.nick = "IQaudIO DAC HiFi pcm512x-hifi-0"
 		media.class = "Audio/Sink"
...
```
It lists all the nodes and you need to find yours. Mine is 58. Copy the Audio/Sink name exactly and it will map mpd's volume to control that sink's volume. 
Pipewire/wireplumber will change node numbers at every boot and the only reliable way to make sure it maps correctly is to use its `node.name`.

Gearhead

Edit. Removed previous patch. I grabbed a patch with debug lines in it. This patch has debug lines stripped out. Built, tested and works. 
[pipewire_volume_1A.patch](https://github.com/user-attachments/files/29887280/pipewire_volume_1A.patch)

**Comment by borine** [comment]
> Let me know if this is more acceptable.

I have no authority to say what is or is not acceptable here, so you are asking the wrong person! I do have one more major concern though, which is the use of `mixer_type`. That is a common setting shared by all audio_output types, so surely it does not make sense to introduce a value that applies only to pipewire?

I think this feature should use `mixer_type "hardware"` and then introduce new setting(s) specific to pipewire, for example:
```
	mixer_type "hardware"
	mixer_target "stream" | "sink"
	mixer_sink "<node.name of the sink>"
```

When `mixer_target` is `"stream"` then the mpd volume is bound to the `channel_volumes` property of the pipewire mpd stream node. This is also the default when `mixer_target` is not given because it is normal application behaviour in pipewire environments.

When `mixer_target` is `"sink"` then if `mixer_sink` is defined the mpd volume is bound to the `channel_volumes` property of the appropriate route of the device associated with the given sink node; or if `mixer_sink` is not defined then it is bound to the `channel_volumes` property of the appropriate route of the default audio sink node as determined by wireplumber.

On paper at least this would appear to work OK for static pipewire graphs where the mpd stream is never moved from its initial target, and also for graphs where the mpd stream is always moved to the default sink by wireplumber whenever that default changes. It fails when the mpd stream is manually linked to a non-default sink by some other client, but I think that is probably OK provided that limitation is mentioned in the mpd user manual.

**Comment by gearhead** [comment]
Interesting thoughts. Since 'stream' volume is how the standard pipewire output controls volume, the 'hardware' option would then pertain to the hardware volume control on either the default sink or a named node sink if it exists.
So if the output type is pipewire, no other parameters is basically how it works, now, volume is the stream volume. 
If the user wants to control the sink volume, they need to specify something and it selects the default sink or a specific node if that is the one that mpd should control.

Since 'stream' is standard volume control, maybe it should be 
mixer_type  "hardware" 
and then if mixer_sink is not populated, it uses "default" and if it is, it uses whichever node is specified? and get rid of 'mixer_target' all together?

I can reconfigure this from a configuration-al perspective however we want it to work. I am trying to provide a similar feature to the alsa hardware volume control, but with pipewire since that was not yet an option. This patch works for me and I can continue to use mpd with this patch, but if there are other comments as to how others may want it to work, I can reconfigure to whatever the consensus is. In the end, I'd like this or some similar functionality in mpd so that we can continue to use it in RuneAudio.

**Comment by borine** [comment]
>  if mixer_sink is not populated, it uses "default" and if it is, it uses whichever node is specified? and get rid of 'mixer_target' all together?

I was thinking that it might be useful to offer 3 alternatives:
* use the stream control
* use the "default" sink control as is done by desktop volume applets and the the ALSA "pipewire" ctl plugin.
* use a specific node given in the mpd configuration (possibly the same as the "target" output, or possibly some other node for whatever reason).

Then a simple "mixer_sink" option is not sufficient, unless you introduce a "special"  string to denote default, such as "%DEFAULT_AUDIO_SINK%" (yuk!)  as used by wpct.

> I can reconfigure to whatever the consensus is

From experience I think you could wait forever to get a consensus here. Take my comments as just random ideas and code up your preference as a PR, and then let Max decide whether it needs to be changed before it can be merged. At least you have got rid of the "exec()" call which was the original showstopper.

**Comment by gearhead** [comment]
Well, the git repo changed slightly and my patch had to be reworked for a variable name change. I decided to take into account some of the points @borine brought up and tried to normalize this a bit with how the config works today. this is what it takes, now:
```
ao {
    type "pipewire"
    # no target, no mixer_type -> default sink, stream volume (today's behavior)
}

ao {
    type        "pipewire"
    mixer_type  "hardware"
    # no target -> plays wherever WirePlumber routes it (default),
    # and volume follows that same default sink, live
}

ao {
    type   "pipewire"
    target "alsa_output.platform-soundcard2.analog-stereo"
    # audio pinned to that card, still stream volume (mirrors plain ALSA "device")
}

ao {
    type        "pipewire"
    target      "alsa_output.platform-soundcard2.analog-stereo"
    mixer_type  "hardware"
    # audio pinned to that card AND volume follows that card's sink
}
```
target is already set-able and if so, this patch makes sure the sink volume controls that sink instead of default. This feels, to me, to align with how mpd is already set up and allows a similar-to-alsa configurability with hardware volume control.

One thing to note is that if the user has configured "hardware" and the default sink changes in wireplumber, this patch does not auto re-route volume control to the new sink. It would require a restart of mpd to do this. Anyone here think that would be a big issue? 

[pipewire_volume_4.patch](https://github.com/user-attachments/files/30846189/pipewire_volume_4.patch)

**Comment by gearhead** [comment]
I tried my hand at editing a couple of the doc files to document a bit what is going on with this patch. The rest is pretty much the same. I confirmed it builds and plays.

This describes what we are doing:
```
This patch adds a new "pwsink" mixer plugin for the PipeWire output. 
Unlike the existing pipewire mixer plugin (PipeWireMixerPlugin.cxx), which
sets SPA_PROP_channelVolumes on MPD's own pw_stream (a per-client 
software gain stage applied before the signal reaches the sink, and not 
shared with other PipeWire clients), this plugin opens its own independent
PipeWire client connection and drives a sink node's volume directly:
 
 - By default it follows the system's current default sink, resolved
   via the "default" pw_metadata object's "default.audio.sink"
   property (the same mechanism @DEFAULT_AUDIO_SINK@ uses).
 - If the output's "target" config option names a specific node, this
   plugin follows that node's volume instead, keeping audio routing
   and volume control pointed at the same sink without a second
   config key.
 
WirePlumber then applies the resulting volume through the ALSA
hardware mixer control when the sink's Route reports one
(route.hw-volume = true), or through PipeWire's own software volume
otherwise -- giving PipeWire outputs the same "real hardware mixer"
option that ALSA, OSS and PulseAudio outputs already have via
mixer_type "hardware", instead of always having a second,
client-side software gain stage on top of whatever WirePlumber itself
applies.
 
MPD's own pw_stream is pinned to unity gain via
pipewire_output_set_volume() while this mixer is open, so the sink's
own volume is the only gain stage in effect.
 
WirePlumber/wpctl display and set volume on a cubic scale rather than
linear PCM gain; this is replicated here (config option
"mixer_volume_curve") so a given MPD volume percentage matches what
"wpctl get-volume" shows for the same node.
```
[pipewire_volume_5.patch](https://github.com/user-attachments/files/30847198/pipewire_volume_5.patch)
